go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,259 @@
+	// SecondaryIndexKey is a part of a secondary index, together with OwnerController and OwnerObjectType
+	// May be used by controllers to store e.g. namespace+name of the object.
+	SecondaryIndexKey = types.OvnK8sPrefix + "/name"
+	// PrimaryIDKey will be used as a primary index, that is unique for every db object,
+	// and can be built based on the combination of all the other ids.
+	PrimaryIDKey = types.OvnK8sPrefix + "/id"
+
+	// Owner Object Types
+	EgressFirewallDNSOwnerType = "EgressFirewallDNS"
+	EgressQoSOwnerType         = "EgressQoS"
+	NetworkPolicyOwnerType     = "NetworkPolicy"
+	NamespaceOwnerType         = "Namespace"
+	HybridNodeRouteOwnerType   = "HybridNodeRoute"
+)
+
+// DbIndex is a structure representing a set of db object ExternalIDs, used to uniquely identify
+// an object in the db (as a primary index), and to filter out a subset of objects based on some IDs
+// (as a secondary index).
+// DbIndex consists of 2 parts:
+// - OwnerController and OwnerObjectType define who manages given object. It is required to communicate with the other
+// entities that create/update/delete db objects.
+// - ObjectIDs uniquely identify objects within the same OwnerController and OwnerObjectType.
+// e.g. if there is only 1 object owned by given controller and objectType in the system, ObjectIDs be empty.
+//
+// Let's consider address set db object that is uniquely identified by namespace and name. It is managed by
+// "controller-1" and that controller uses objectType "AddressSetType" for those objects. We can define
+// addrSetIDs struct to work with these ids, for that it needs to implement DbObjectIDs interface.
+//
+//	const (
+//		ipFamilyKey = "ipFamily"
+//		Controller = "controller-1"
+//		ObjectType = "AddressSetType"
+//	)
+//
+//	type addrSetIDs struct {
+//		NamespaceName string
+//      ipFamily string
+//	}
+//
+//	func (a *addrSetIDs) GetExternalIDs() map[string]string {
+//		return map[string]string{
+//			SecondaryIndexKey: a.NamespaceName,
+//			ipFamilyKey:      a.ipFamily,
+//		}
+//	}
+//	func (a *addrSetIDs) GetUniqueID() string {
+//		uniqueID := a.NamespaceName
+//		if a.ipFamily != "" {
+//			uniqueID += ":" + a.ipFamily
+//		}
+//		return uniqueID
+//	}
+//
+// This struct may be used to create DbIndex
+//	dbIndex := NewDbIndex[*addrSetIDs](Controller, ObjectType, &addrSetIDs{
+//		NamespaceName: "namespace1_name1",
+//		ipFamily: "ipv4",
+//	})
+//
+// This index may be used to create db object
+//	as := &nbdb.AddressSet{
+//		Name:        dbIndex.GetUniqueID(),
+//		ExternalIDs: dbIndex.GetExternalIDs(),
+//	}
+//
+// The following ExternalIDs will be created for this example
+// {
+//		"k8s.ovn.org/controller": "controller-1"
+//		"k8s.ovn.org/object-type": "AddressSetType"
+//		"k8s.ovn.org/name": "namespace1_name1"
+//		"k8s.ovn.org/id": "AddressSetType.controller-1.k8s.ovn.org/namespace1_name1:ipv4"
+//		"AddressSetType.controller-1.k8s.ovn.org/ipFamily": "ipv4"
```

Comment:
What would we use this external id for? Only within a predicate? Or as a client index? I don't think we could use something as variable as "controller-1" as part of an external id key which is part of a client index. Something less variable like "AddressSetType" could probably be accommodated with further minor changes in libovsdb. The problem is that if not all rows of the table have this external id key, we would be tracking indexes with default values for no purpose which would be inefficient.

Reply:
in that version, all object-specific ids (that have <objecttype> prefix) can only be used for predicate search, or for anything controller needs after the object is fetched from the db.

Reply:
My question would then be why would the handle these types of IDs generically here (unless there are plans to distinguish here between IDs to be used as secondary indexes vs IDs to be used in predicate only).

But now that I look at it, maybe we aren't? I don't know where `"AddressSetType.controller-1.k8s.ovn.org/ipFamily": "ipv4"` is build, at least I don't see it in this example.
I see `AddressSetType.controller-1.k8s.ovn.org` is the owner key but not the `/ipFamily": "ipv4` part.